### PR TITLE
Remove Widget, add post-render hooks

### PR DESCRIPTION
**Pull request by paf31**
/cc @jdegoes @cryogenian @garyb @puffnfresh

This PR is a bit more drastic. @garyb and I were discussing widgets, and this demonstrates an idea we were considering - since widgets are effectful, instead of trying to shoehorn them into the pure world of `Component`, it makes sense to provide a low-level, imperative escape hatch to deal with them. It's not pretty, but widgets aren't either, and it fits with the "push effects to the application boundary" approach. 

Namely, I've removed `Widget`, and added support for a "post-render hook" to `runUIWith`. Widgets can be synchronized in the post-render method after every patch is rendered.

The Ace example shows an example of using `data-` attributes to synchronize the Ace widget state with the state of the component. 

@jdegoes I think this is more in line with the "big record of widget state" approach that we discussed in #84.

### CODE REVIEW

**Comment by paf31** [comment]
Also, if we go this route, I don't think we need to have a separate `halogen-echarts` repo, since we will just be using the ECharts API directly.

**Comment by paf31** [comment]
Resolves #98, since we can create `data-` attributes to represent the focus state.
Resolves #100.

File: docs/Halogen-HTML-Renderer.md [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #### `renderHTML`
 
 ``` purescript
-renderHTML :: forall p i eff. (i -> Eff eff Unit) -> (p -> Widget eff i) -> H.HTML p i -> VTree
+renderHTML :: forall i eff. (i -> Eff eff Unit) -> H.HTML Void i -> VTree
```
Comment (puffnfresh) [inline docs/Halogen-HTML-Renderer.md]: :+1:

File: docs/Halogen.md [hunk 2]
```diff
@@ -60,7 +60,7 @@ main = do
 #### `runUI`
 
 ``` purescript
-runUI :: forall req eff. Component (Widget (HalogenEffects eff) req) (Event (HalogenEffects eff)) req req -> Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
+runUI :: forall req eff. Component Void (Event (HalogenEffects eff)) req req -> Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
```
Comment (puffnfresh) [inline docs/Halogen.md]: Nice! :+1:

**Comment by puffnfresh** [comment]
Brilliant. I think these updates will unblock me with integration of Ace into SlamData. Thank you!

File: README.md [hunk 3]
```diff
@@ -100,17 +100,15 @@ See the AJAX example for a demonstration of the `Event` monad.
...
 
-In practice, Halogen chooses `p` to be the type `Widget eff input`, which describes a `virtual-dom` widget, but as with the type argument `m`, leaving things polymorphic has the advantage that we can choose to interpret our model in a different way for testing.
+The `runUIWith` function accepts an extra parameter, a function which will be called whenever the DOM is updated. By using this function in conjunction with custom data attributes, it is possible to synchronize the internal state of a third-party component with your component's internal state.
 
-The `p` type parameter also gives us a way to "graft" HTML documents at a point which is marked by a placeholder. We can think of the type `HTML p a` as describing HTML documents with "holes" marked with values of type `p`.
+See the `examples/ace` directory for a worked example.
 
-See the Ace editor example for a demonstration of placeholders.
+Note: this is considered an advanced use case - incorrect use of this feature could lead to unexpected behavior, so it is recommended that you understand the rendering pipeline before using post-render hooks.
```
Comment (jdegoes) [inline README.md]: This comment makes me sad. :frowning: Is the interface as safe as can be, or are there additional ways we can improve its safety?

**Comment by jdegoes** [comment]
> @jdegoes I think this is more in line with the "big record of widget state" approach that we discussed in #84.

I can't help but feel this approach will implode at some point in the very near term (Ace with a large enough buffer, a Canvas-based paint component, a chart with a lot of data points, etc.). But overall I think these changes are a positive direction from where we are right now, so in the spirit of making incremental progress. :+1: .

**Comment by jdegoes** [comment]
@paf31 The other thing: there's no need for `p` anymore, is it time to remove that to simplify `HTML` and `Component`?

**Comment by paf31** [comment]
My original PR also removed `p`, but I decided to change things in stages because it broke the downstream libraries, and I also wasn't sure if we wanted to keep `install` etc.

If we want to remove `p`, I can work on that today.

**Comment by garyb** [comment]
:+1: for removing `p`.

**Pull request merged**